domain/credential/service/service.go

```diff
@@ -50,31 +57,90 @@ type State interface {
 		getWatcher func(string, string, changestream.ChangeType) (watcher.NotifyWatcher, error),
 		name, cloudName, owner string,
 	) (watcher.NotifyWatcher, error)
+
+	// ModelsUsingCloudCredential returns a map of uuid->name for models which use the credential.
+	ModelsUsingCloudCredential(ctx context.Context, id credential.ID) (map[model.UUID]string, error)
+
+	// GetCloud returns the cloud with the specified name.
+	GetCloud(ctx context.Context, name string) (cloud.Cloud, error)
 }
 
+// ValidationContextGetter returns the artefacts for a specified model, used to make credential validation calls.
+type ValidationContextGetter func(modelUUID model.UUID) (credentialcommon.CredentialValidationContext, error)
+
 // Logger facilitates emitting log messages.
 type Logger interface {
-	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
 }
 
 // Service provides the API for working with external controllers.
 type Service struct {
 	st             State
 	watcherFactory WatcherFactory
+	logger         Logger
+
+	// These are set via options after the service is created.
+
+	validationContextGetter ValidationContextGetter
+	validator               credentialcommon.CredentialValidator
+
+	// TODO(wallyworld) - remove when models are out of mongo
+	legacyUpdater func(tag names.CloudCredentialTag) error
+	legacyRemover func(tag names.CloudCredentialTag) error
 }
 
 // NewService returns a new service reference wrapping the input state.
-func NewService(st State, watcherFactory WatcherFactory) *Service {
+func NewService(st State, watcherFactory WatcherFactory, logger Logger) *Service {
 	return &Service{
 		st:             st,
 		watcherFactory: watcherFactory,
+		logger:         logger,
+		validator:      noopValidator{},
 	}
 }
 
-// CloudCredential represents a credential and the cloud it belongs to.
-type CloudCredential struct {
-	Credential cloud.Credential
-	CloudName  string
+// noopValidator always validates any credential.
+// TODO(wallyworld) - this is a placeholder.
+type noopValidator struct {
+}
+
+func (noopValidator) Validate(
+	_ credentialcommon.CredentialValidationContext,
+	_ credential.ID,
+	_ *cloud.Credential,
+	_ bool,
+) (machineErrors []error, err error) {
+	return nil, nil
+}
+
+// WithValidationContextGetter configures the service to use the specified function
+// to get a context used to validate a credential for a specified model.
+func (s *Service) WithValidationContextGetter(validationContextGetter ValidationContextGetter) *Service {
```

Comment:
Yes, validation uses state machine objects and environ instances etc. We need to avoid introducing that rats nest of dependencies here.

Reply:
Instead of this design where we are playing Russian roulette with some what constructed state why not keep the same pattern but have a sub service returned that can do the actual update so we are always playing with objects that are in a valid state?

Example of this would be:
```
func (s *Service) NewUpdatingCredentialService(//prereqs ....) *UpdatingService
```

You can only then do updates on a service that has everything and you can only get one off of a credential service and it can share it's internal's.

I am just not a fan of design where the user needs to check again that a constructed object is in a good state to use it. If you have it, it should be good to rock and roll.

Reply:
The user doesn't need to check anything. The implementation checks if the context getter is nil when the Update method is called and returns an error if needed.
With both approaches, you still need to set it up correctly via an extra call, either service.WithFoo() or service.NewUpdatingService(). And the approach I've used conforms to the pattern we've used elsewhere in dqlite and other places, and allows you to set up the api facade to use the one service everywhere.

Reply:
As discussed, I've added extra comments to document the expectations. The method should be able to be removed once on dqlite fully (or refactored, TBD).